iroh-gossip/src/proto/hyparview.rs

```diff
@@ -0,0 +1,707 @@
+            rng,
+            stats: Stats::default(),
+            pending_neighbor_requests: Default::default(),
+            peer_data: Default::default(),
+        }
+    }
+
+    pub fn handle(&mut self, event: InEvent<PA>, now: Instant, io: &mut impl IO<PA>) {
+        match event {
+            InEvent::RecvMessage(from, message) => self.handle_message(from, message, now, io),
+            InEvent::TimerExpired(timer) => match timer {
+                Timer::DoShuffle => self.handle_shuffle_timer(io),
+                Timer::PendingNeighborRequest(peer) => self.handle_pending_neighbor_timer(peer, io),
+            },
+            InEvent::PeerDisconnected(peer) => self.handle_disconnect(peer, io),
+            InEvent::RequestJoin(peer) => self.handle_join(peer, io),
+            InEvent::UpdatePeerData(data) => {
+                self.me_data = data;
+            }
+            InEvent::Quit => self.handle_quit(io),
+        }
+
+        // this will only happen on the first call
+        if !self.shuffle_scheduled {
+            io.push(OutEvent::ScheduleTimer(
+                self.config.shuffle_interval,
+                Timer::DoShuffle,
+            ));
+            self.shuffle_scheduled = true;
+        }
+    }
+
+    fn handle_message(
+        &mut self,
+        from: PA,
+        message: Message<PA>,
+        now: Instant,
+        io: &mut impl IO<PA>,
+    ) {
+        let is_disconnect = matches!(message, Message::Disconnect(Disconnect { .. }));
+        if !is_disconnect && !self.active_view.contains(&from) {
+            self.stats.total_connections += 1;
+        }
+        match message {
+            Message::Join(data) => self.on_join(from, data, now, io),
+            Message::ForwardJoin(details) => self.on_forward_join(from, details, now, io),
+            Message::Shuffle(details) => self.on_shuffle(from, details, io),
+            Message::ShuffleReply(details) => self.on_shuffle_reply(details, io),
+            Message::Neighbor(details) => self.on_neighbor(from, details, now, io),
+            Message::Disconnect(details) => self.on_disconnect(from, details, io),
+        }
+
+        // Disconnect from passive nodes right after receiving a message.
+        if !is_disconnect && !self.active_view.contains(&from) {
+            io.push(OutEvent::DisconnectPeer(from));
+        }
+    }
+
+    fn handle_join(&mut self, peer: PA, io: &mut impl IO<PA>) {
+        io.push(OutEvent::SendMessage(
+            peer,
+            Message::Join(self.me_data.clone()),
+        ));
+    }
+
+    fn handle_disconnect(&mut self, peer: PA, io: &mut impl IO<PA>) {
+        self.on_disconnect(
+            peer,
+            Disconnect {
+                alive: true,
+                respond: false,
+            },
+            io,
+        );
+    }
+
+    fn handle_quit(&mut self, io: &mut impl IO<PA>) {
+        let peers = self.active_view.iter().cloned().collect::<Vec<_>>();
+        for peer in peers {
```

Comment:
```suggestion
        for peer in self.active_view.clone().into_iter() {
```
I added the `IntoIter` implementation in another suggestion. Saves you the collect